File: src/main/java/hu/elte/alkfejl/orarend/model/Course.java
package hu.elte.alkfejl.orarend.model;

import hu.elte.alkfejl.orarend.service.schedule.Interval;
import hu.elte.alkfejl.orarend.service.schedule.IntervalConverter;
import lombok.*;

import javax.persistence.*;

@Entity
@Table(name="COURSES")
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class Course extends BaseEntity {

    @Column(nullable = false)
    private String name; //0

    @Column(nullable = false)
    private String subCode; //1

    @Column(nullable = false)
    @Convert(converter = IntervalConverter.class)
    private Interval interval; //2

    @Column(nullable = false)
    private String room; //3

    @Column(nullable = false)
    private String note; //5

    @Column(nullable = true)
    @Enumerated(EnumType.STRING)
    private CourseType type; //6

    @Column(nullable = false)
    private int courseCode; //7

    @Column(nullable = false)
    private String teacher; //11

    boolean overlapsWith(Course other) {
        return interval.overlapsWith(other.getInterval());
    }

    @Override
    public String toString() {
        return "Course{" +
                "name='" + name + '\'' +
                ", subCode='" + subCode + '\'' +
                ", interval=" + interval +
                ", room='" + room + '\'' +
                ", note='" + note + '\'' +
                ", type=" + type +
                ", courseCode=" + courseCode +
                ", teacher='" + teacher + '\'' +
                '}';
    }

    public enum CourseType {
        Practice, Presentation, Consultation, Unknown;

        @Override
        public String toString() {
            switch(this) {
                case Practice:
                    return "Practice";
                case Presentation:
                    return "Presentation";
                case Consultation:
                    return "Consultation";
                default:
                    return "";
            }
        }
    }

}

File: orarend-tervezo/src/app/model/Course.ts
// "{
// "day":"MONDAY",
// "endTime":{"hour":14,"min":0},
// "startTime":{"hour":12,"min":0},
// "length":120}"
export class Time {
  private hour: number;
  private min: number;

  constructor(hour: number, min: number) {
    this.hour = hour;
    this.min = min;
  }
}

export class Interval {
  private day: String;
  private startTime: Time;
  private endTime: Time;
  private length: number;

  constructor(day?: String, startTime?: Time, endTime?: Time, length?: number) {
    this.day = day;
    this.startTime = startTime;
    this.endTime = endTime;
    this.length = length;
  }
}

export class Course {
  private id: number;
  private version: number;
  private name: String;
  private subCode: String;
  private interval: Interval;
  private room: String;
  private note: String;
  private type: String;
  private courseCode: number;
  private teacher: String;

  constructor(id?: number, version?: number, name?: String, subCode?: String, interval?: Interval, room?: String, note?: String, type?: String, courseCode?: number, teacher?: String) {
    this.id = id;
    this.version = version;
    this.name = name;
    this.subCode = subCode;
    this.interval = interval;
    this.room = room;
    this.note = note;
    this.type = type;
    this.courseCode = courseCode;
    this.teacher = teacher;
  }
}
